### PR TITLE
Add permanent storage option for EventStream

### PR DESCRIPTION
**Pull request by rbren**
This allows EventStream to persist to a local or remote filesystem, instead of keeping all state in-memory.

This is helpful for a high-availability situation. For example, currently, if you rolled out a new deployment, any websockets would be disconnected, existing user sessions would be cancelled, and all the session data would be lost.

With this change, the client could reconnect to a new instance of the application, and the EventStore could rehydrate from persistent storage.

This change also factors action/observation/event serialization and deserialization into its own spot, to avoid circular imports

### CODE REVIEW

File: opendevin/storage/s3.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import os
+
+from minio import Minio
+
+from .files import FileStore
+
+AWS_S3_ENDPOINT = 's3.amazonaws.com'
+
+
+class S3FileStore(FileStore):
```
Comment (assertion) [inline opendevin/storage/s3.py:10]: I think maybe put s3.py into a directory of cloud provider will be better, so we can add some other cloud file storage like azure bolb, Google Cloud Storage.

File: opendevin/events/stream.py [hunk 2]
```diff
@@ -21,9 +25,32 @@ class EventSource(str, Enum):
...
+            self._events = []
+            events = self._file_store.list(f'sessions/{self.sid}/events')
+            print('list of events', events)
+            for event_str in events:
+                content = self._file_store.read(event_str)
+                data = json.loads(content)
+                event = event_from_dict(data)
+                self._events.append(event)
 
     def subscribe(self, id: EventStreamSubscriber, callback: Callable):
```
Comment (assertion) [inline opendevin/events/stream.py:55]: Hi Robert, the current subscribe/unsubscribe strategy is based on `agent_controller`, `server`, `runtime`, `main` enum. It's not based on the instance. 
So  one OpenDevin server can't serve two sessions(two users or two browsers chat with opendevin) at the same time right? The second subscribe call will be rejected ('Subscriber subscribed multiple times: ' + id) .

**Comment by assertion** [comment]
@rbren I found this bug when opening two browsers and chat with opendevin at the same time, the bug related to EventStream. 
pr: https://github.com/OpenDevin/OpenDevin/pull/1713

And I found you fixed it in current pr, but I suggest to merge mine pr before you finish current one.

**Comment by rbren** [comment]
good catch @assertion!

**Review by enyst — APPROVED** [review]

**Comment by enyst** [comment]
I haven't got to test it, otherwise LGTM!

File: opendevin/events/stream.py [hunk 3]
```diff
@@ -41,9 +67,13 @@ def unsubscribe(self, id: EventStreamSubscriber):
...
-            event._source = source  # type: ignore[attr-defined]
+            event._id = len(self._events)  # type: ignore [attr-defined]
+            event._timestamp = datetime.now()  # type: ignore [attr-defined]
+            event._source = source  # type: ignore [attr-defined]
+            self._file_store.write(
+                self._get_filename_for_event(event), event_to_json(event)
+            )
             self._events.append(event)
         for key, fn in self._subscribers.items():
+            print('calling subscriber', key)
```
Comment (rbren) [inline opendevin/events/stream.py:78]:
```suggestion
```

File: opendevin/storage/s3.py (hunk 1 above)
Comment (rbren) [reply opendevin/storage/s3.py:10]: agree we'll probably add these in the future! Let's create a subdirectory then if we decide we need it

**Pull request merged**